Arrowgene.Ddon.GameServer/Characters/WeatherManager.cs

```diff
@@ -0,0 +1,166 @@
+
+        private static readonly long OriginalRealTimeSec=0x55DDD470; // Taken from the pcaps. A few days before DDOn release
+        private static readonly long OriginalGameTimeSec=0x22C2ED000; // Taken from the pcaps.
+        private static readonly uint GameTimeDayHour = 24;
+
+        public static readonly uint ForecastIntervalGameHour = 6; //Used in 
+
+        private readonly uint GameClockTimescale;
+
+        private readonly DdonGameServer _Server;
+
+        public List<CDataWeatherLoop> WeatherLoopList { get; private set; }
+        public ulong WeatherLoopTotalLength { get; private set; }
+
+        public WeatherManager(DdonGameServer server)
+        {
+            _Server = server;
+            GameClockTimescale = server.Setting.GameLogicSetting.GameClockTimescale;
+            GenerateWeatherSequence();
+        }
+
+        public Weather GetWeather()
+        {
+            return GetWeather(DateTimeOffset.UtcNow);
+        }
+
+        public Weather GetWeather(DateTimeOffset time)
+        {
+            ulong secondsElapsed = (ulong)(time.ToUnixTimeSeconds() - OriginalRealTimeSec);
+            ulong remainingSeconds = secondsElapsed % WeatherLoopTotalLength;
+            foreach (var weatherLoop in WeatherLoopList)
+            {
+                if (remainingSeconds <= weatherLoop.TimeSec)
+                {
+                    return weatherLoop.WeatherId;
+                }
+                remainingSeconds -= weatherLoop.TimeSec;
+            }
+            return Weather.Fine;
+        }
+
+        public List<CDataWeatherForecast> GetForecast()
+        {
+            List<CDataWeatherForecast> forecast = new List<CDataWeatherForecast>();
+
+            // The first entry on the forecast is always the next Lestanian hour.
+            // Weather changes are set up on the hour, so we go one second over to see what's coming up.
+            DateTimeOffset nextHour = GetTimeForNextGameHour().AddSeconds(1);
+            forecast.Add(new CDataWeatherForecast()
+            {
+                Weather = GetWeather(nextHour)
+            });
+
+            uint forecastInterval = ServerWeatherForecastGetHandler.IntervalGameHour;
+            DateTimeOffset forecastTime = nextHour;
+            for (int i = 0; i < 3; i++)
+            {
+                forecastTime = forecastTime.AddMinutes(GameClockTimescale * forecastInterval / GameTimeDayHour);
+                forecast.Add(new CDataWeatherForecast()
+                {
+                    Weather = GetWeather(forecastTime)
+                });
+            }
+
+            return forecast;
+        }
+
+        private DateTimeOffset GetTimeForNextGameHour()
+        {
+            DateTimeOffset now = DateTimeOffset.UtcNow;
+            long currentTimeMS = RealTimeToGameTimeMS(now);
+            double currentTimeHour = currentTimeMS / 3600000.0;
+            uint offsetSecondsToNextHour = (uint)((1 - currentTimeHour % 1) * 3600);
```

Comment:
I don't understand this math. Won't this underflow? Current hour is 1-24 (or 0-23)? So If the hour was 3.

1 - 3 = -2. This is also a modulus operator on a double which is negative? Please help clarify :)

Reply:
Can you add parenthesis to make the order of operations clearer.